### ISSUE TITLE
Using CuDNN results in worse performance

### ISSUE THREAD

**Issue by sachag678** [deeplearning4j/deeplearning4j#9376]
### Issue Description

I'm doing some supervised training using a ConvNet. I've attached the network design.

#### Test case 1
In the first case my sbt has the following setup for dl4j:

11:53:12.801 [main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [JCublasBackend] backend
11:53:13.838 [main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for linear algebra: 32
11:53:13.863 [main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CUDA]; OS: [Linux]
11:53:13.863 [main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [12]; Memory: [3.5GB];
11:53:13.863 [main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [CUBLAS]
11:53:13.868 [main] INFO org.nd4j.linalg.jcublas.JCublasBackend - ND4J CUDA build version: 10.2.89
11:53:13.869 [main] INFO org.nd4j.linalg.jcublas.JCublasBackend - CUDA device 0: [GeForce GTX 960]; cc: [5.2]; Total memory: [2098069504]
11:53:13.899 [main] INFO org.deeplearning4j.nn.multilayer.MultiLayerNetwork - Starting MultiLayerNetwork with WorkspaceModes set to [training: ENABLED; inference: ENABLED], cacheMode set to [NONE]
11:53:14.271 [main] INFO org.deeplearning4j.nn.layers.convolution.ConvolutionLayer - cuDNN not found: use cuDNN for better GPU performance by including the deeplearning4j-cuda module. For more information, please refer to: https://deeplearning4j.konduit.ai/config/backends/config-cudnn
java.lang.ClassNotFoundException: org.deeplearning4j.cuda.convolution.CudnnConvolutionHelper
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at org.deeplearning4j.nn.layers.convolution.ConvolutionLayer.initializeHelper(ConvolutionLayer.java:77)
	at org.deeplearning4j.nn.layers.convolution.ConvolutionLayer.<init>(ConvolutionLayer.java:69)
	at org.deeplearning4j.nn.conf.layers.ConvolutionLayer.instantiate(ConvolutionLayer.java:171)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.init(MultiLayerNetwork.java:719)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.init(MultiLayerNetwork.java:609)
	at Trainer$.trainConvNet(Trainer.scala:170)
	at Trainer$.main(Trainer.scala:391)
	at Trainer.main(Trainer.scala)

    libraryDependencies += "com.typesafe.scala-logging" %% "scala-logging" % "3.4.0",
    libraryDependencies += "ch.qos.logback" % "logback-classic" % "1.2.3",
    libraryDependencies += "org.deeplearning4j" % "deeplearning4j-core" % "1.0.0-beta7",
    libraryDependencies += "org.nd4j" % "nd4j-cuda-10.2" % "1.0.0-beta7",
    libraryDependencies += "org.deeplearning4j" % "arbiter-deeplearning4j" % "1.0.0-beta7",
    libraryDependencies += "org.deeplearning4j" % "arbiter-ui" % "1.0.0-beta7",
    libraryDependencies += "org.deeplearning4j" % "deeplearning4j-ui" % "1.0.0-beta7",
    libraryDependencies += "org.deeplearning4j" % "arbiter-core" % "1.0.0-beta7"

Note using deeplearning-core 1.0.0-beta7 (and as such it says no cuDNN found in the log)

I run my own training loop where I format the data and fit the network and then test on an unseen test set. I calculate the MAE manually and print it out as shown below.

Starting training at (Thu Jun 17 11:53:14 EDT 2021)
Epoch 0 => Score(MAE): 0.06335164115856526 (Thu Jun 17 15:23:46 EDT 2021)
Epoch 1 => Score(MAE): 0.05033815414917979 (Thu Jun 17 18:54:03 EDT 2021)
Epoch 2 => Score(MAE): 0.04600503341433101 (Thu Jun 17 22:22:58 EDT 2021)
Epoch 3 => Score(MAE): 0.038528926136972 (Fri Jun 18 01:50:56 EDT 2021)
Reached Max epoch limit
Best Score: 0.038528926136972 at Epoch 3.0

#### Test case 2

09:31:39.673 [main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [JCublasBackend] backend
09:31:41.079 [main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for linear algebra: 32
09:31:41.107 [main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CUDA]; OS: [Linux]
09:31:41.107 [main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [12]; Memory: [3.5GB];
09:31:41.107 [main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [CUBLAS]
09:31:41.115 [main] INFO org.nd4j.linalg.jcublas.JCublasBackend - ND4J CUDA build version: 10.2.89
09:31:41.117 [main] INFO org.nd4j.linalg.jcublas.JCublasBackend - CUDA device 0: [GeForce GTX 960]; cc: [5.2]; Total memory: [2098069504]
09:31:41.157 [main] INFO org.deeplearning4j.nn.multilayer.MultiLayerNetwork - Starting MultiLayerNetwork with WorkspaceModes set to [training: ENABLED; inference: ENABLED], cacheMode set to [NONE]
09:31:42.790 [main] DEBUG org.deeplearning4j.nn.layers.convolution.ConvolutionLayer - CudnnConvolutionHelper successfully initialized
09:31:42.808 [main] DEBUG org.deeplearning4j.nn.layers.convolution.ConvolutionLayer - CudnnConvolutionHelper successfully initialized
09:31:42.813 [main] DEBUG org.deeplearning4j.nn.layers.convolution.ConvolutionLayer - CudnnConvolutionHelper successfully initialized
09:31:42.816 [main] DEBUG org.deeplearning4j.nn.layers.convolution.ConvolutionLayer - CudnnConvolutionHelper successfully initialized
09:31:42.818 [main] DEBUG org.deeplearning4j.nn.layers.convolution.ConvolutionLayer - CudnnConvolutionHelper successfully initialized
09:31:42.822 [main] DEBUG org.deeplearning4j.nn.layers.convolution.subsampling.SubsamplingLayer - CudnnSubsamplingHelper successfully initialized

 libraryDependencies += "com.typesafe.scala-logging" %% "scala-logging" % "3.4.0",
    libraryDependencies += "ch.qos.logback" % "logback-classic" % "1.2.3",
    libraryDependencies += "org.deeplearning4j" % "deeplearning4j-cuda-10.2" % "1.0.0-beta7",
    libraryDependencies += "org.nd4j" % "nd4j-cuda-10.2" % "1.0.0-beta7",
    libraryDependencies += "org.deeplearning4j" % "arbiter-deeplearning4j" % "1.0.0-beta7",
    libraryDependencies += "org.deeplearning4j" % "arbiter-ui" % "1.0.0-beta7",
    libraryDependencies += "org.deeplearning4j" % "deeplearning4j-ui" % "1.0.0-beta7",
    libraryDependencies += "org.deeplearning4j" % "arbiter-core" % "1.0.0-beta7"

Now using deeplearning4j-cuda-10.2 (and as such shows in the log that it found the CudnnHelpers)

Starting training at (Thu Jun 17 09:31:42 EDT 2021)
Epoch 0 => Score(MAE): 0.05368905329361064 (Thu Jun 17 10:05:41 EDT 2021)
Epoch 1 => Score(MAE): 0.04705279847639493 (Thu Jun 17 10:39:38 EDT 2021)
Epoch 2 => Score(MAE): 0.06169708284285256 (Thu Jun 17 11:13:35 EDT 2021)
Epoch 3 => Score(MAE): 0.05375517608196155 (Thu Jun 17 11:48:01 EDT 2021)
Reached Max epoch limit
Best Score: 0.04705279847639493 at Epoch 1.0

The results are worse and much more erratic.

#### Questions

Test Case 2 is much faster almost 4x. 
Is this expected behavior to sacrifice stability and have increased error for speed?

The NN are used as State-value networks in a game and Test Case 1's network performs better than Test Case 2s.

#### Version Information
Using Ubuntu 18.04.5 LTS

Cuda-10.2 as stated above

Just installed libcudnn8_8.2.1.32-1+cuda10.2_amd64.deb for cuDNN.

+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.119.03   Driver Version: 450.119.03   CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GeForce GTX 960     Off  | 00000000:26:00.0  On |                  N/A |
|  0%   52C    P8     9W / 160W |   1580MiB /  2000MiB |      1%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      1526      G   /usr/lib/xorg/Xorg                 39MiB |
|    0   N/A  N/A      1661      G   /usr/bin/gnome-shell               47MiB |
|    0   N/A  N/A      2624      G   /usr/lib/xorg/Xorg                556MiB |
|    0   N/A  N/A      2786      G   /usr/bin/gnome-shell              189MiB |
|    0   N/A  N/A      3138      G   ...C7E2C5142C3B28EEEC86259F1       83MiB |
|    0   N/A  N/A      3547      G   ...AAAAAAAAA= --shared-files      370MiB |
|    0   N/A  N/A     28986      G   /usr/lib/firefox/firefox          280MiB |
+-----------------------------------------------------------------------------+

#### Files

[NNDesign.txt](https://github.com/eclipse/deeplearning4j/files/6677626/NNDesign.txt)

**Comment by agibsonccc** [comment]
@sachag678 I'll have to defer this to the next version maybe...a lot of cudnn changes are coming. I would ask that you re open once the newer version is out next week. If you want to try using cudnn, you can look here: https://deeplearning4j.konduit.ai/v/en-1.0.0-m1/multi-project/explanation/configuration/backends/cudnn  - you may also try either using M1 or snapshots: https://deeplearning4j.konduit.ai/v/en-1.0.0-m1/multi-project/explanation/snapshots W

**Comment by sachag678** [comment]
Okay I will try using the M1 version and see if the same issue persists. Thanks!

**Comment by sachag678** [comment]
I also just wanted to note - I tried the same experiment but with a slightly different network design. There was the opposite behavior where the cuDNN performed better so it seems that it is a case by case basis.
